### ISSUE TITLE
Roslyn ignores `[MemberNotNull]` annotations on implicitly called `GetEnumerator` methods

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#50038]
**Version Used**: 
```
Branch master (17 Dec 2020)
Latest commit 4511a45 by msftbot[bot]:
Merge pull request #49967 from mavasani/CodeStyleConfigFileGenerator

Add tooling to CodeStyle layer to generate global configs
```

**Steps to Reproduce**:

Compile the following code
```cs
using System.Collections;
using System.Diagnostics.CodeAnalysis;

#nullable enable
class C
{
  public string? Field;

  void M1(C c)
  {
    foreach (var x in c)
    {
      // false CS8602 warning after GetEnumerator annotated with [MemberNotNull("Field")] was called
      c.Field.ToString();
    }
  }
    
  void M2(C c)
  {
    // if you call it explicitly there are no warnings
    IEnumerator enumerator = c.GetEnumerator();
    while (enumerator.MoveNext())
    {
      object? current = enumerator.Current;
      c.Field.ToString();
    }
  }
  
  [MemberNotNull("Field")] public extern IEnumerator GetEnumerator();
}
```

**Expected Behavior**:
No warnings. `foreach (var x in c)` has called `c.GetEnumerator` method that is annotated with `[MemberNotNull("Field")]` therefore `c.Field` must be not nullable

**Actual Behavior**:
False `warning CS8602: Dereference of a possibly null reference.` is reported in `M1`.
Note that there is no such warning if you rewrite the `foreach` loop with explicit `GetEnumerator` call
